File: src/Queue/OrderedLinkedQueue.java
package Queue;

import java.util.Comparator;

public class OrderedLinkedQueue<T> extends LinkedQueue<T> {
    private Comparator<T> comparator;

    public OrderedLinkedQueue(Comparator<T> comparator) {
        super();
        this.comparator = comparator;
    }

    @Override
    public void add(T element) {
        var p = store.getHeadNode();
        while (p.getNext() != null) {
            if (comparator.compare(element, p.getNext().getData()) > 0) {
                p = p.getNext();
            } else {
                p.setNext(store.getNewNode(element, p.getNext()));
                return;
            }
        }
        store.insertTail(element);
    }
}

File: src/Queue/OrderedLinkedQueue.kt
package Queue

class OrderedLinkedQueue<T>(
        private val comparator: Comparator<T>
) : LinkedQueue<T>() {

    override fun add(element: T) {
        var p = store.headNode
        while (p.next != null) {
            if (comparator.compare(element, p.next.data) > 0) {
                p = p.next
            } else {
                p.next = store.getNewNode(element, p.next)
                return
            }
        }
        store.insertTail(element)
    }
}
